nexus/src/app/background/tasks/instance_reincarnation.rs

```diff
@@ -0,0 +1,715 @@
+
+//! Background task for automatically restarting failed instances.
+
+use crate::app::background::BackgroundTask;
+use crate::app::saga::StartSaga;
+use crate::app::sagas::instance_start;
+use crate::app::sagas::NexusSaga;
+use chrono::TimeDelta;
+use chrono::Utc;
+use futures::future::BoxFuture;
+use nexus_db_queries::authn;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::DataStore;
+use nexus_types::identity::Resource;
+use nexus_types::internal_api::background::InstanceReincarnationStatus;
+use omicron_common::api::external::Error;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::task::JoinSet;
+
+pub struct InstanceReincarnation {
+    datastore: Arc<DataStore>,
+    sagas: Arc<dyn StartSaga>,
+    /// The default cooldown period between automatic restarts.
+    ///
+    /// If an instance's last automatic restart occurred less than this duration
+    /// from now, it may not be automatically restarted until it's had some time
+    /// to calm down. This is intended to try and reduce the impact of tight
+    /// crash loops.
+    //
+    // TODO(eliza): this default should be overridden by a project-level default
+    // when https://github.com/oxidecomputer/omicron/issues/1015 is implemented.
+    default_cooldown: TimeDelta,
+}
+
+impl BackgroundTask for InstanceReincarnation {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        Box::pin(async move {
+            let mut status = InstanceReincarnationStatus::default();
+            status.default_cooldown = self.default_cooldown.to_std().expect(
+                "cooldown came from a `std::time::Duration` which was \
+                 non-negative, so it should remain non-negative",
+            );
+            self.actually_activate(opctx, &mut status).await;
+            if !status.restart_errors.is_empty() || status.query_error.is_some()
+            {
+                error!(
+                    &opctx.log,
+                    "instance reincarnation completed with errors!";
+                    "instances_found" => status.instances_found,
+                    "instances_reincarnated" => status.instances_reincarnated.len(),
+                    "already_reincarnated" => status.already_reincarnated.len(),
+                    "query_error" => ?status.query_error,
+                    "restart_errors" => status.restart_errors.len(),
+                );
+            } else if !status.instances_reincarnated.is_empty() {
+                info!(
+                    &opctx.log,
+                    "instance reincarnation completed";
+                    "instances_found" => status.instances_found,
+                    "instances_reincarnated" => status.instances_reincarnated.len(),
+                    "already_reincarnated" => status.already_reincarnated.len(),
+                );
+            } else {
+                debug!(
```

Comment:
small: I'd consider always logging this at "info".  We're not doing this very often I think and it may be useful for putting together a timeline of events if this ever doesn't do what we think it should have done.

Reply:
sounds good! perhaps the log record for reincarnating an individual instance should also be bumped up to `info` --- I'll go do that.